xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,409 @@
+	// pending balancers are both populated), this error will always be
+	// forwarded to the pending. Thus, there is a guarantee that this will not
+	// break the balancer API of the balancer by updating after closing.
+	balToUpdate.ResolverError(err)
+}
+
+func (gsb *gracefulSwitchBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	// At any given time, the current balancer may be closed in the situation
+	// where there is a current balancer and a pending balancer. Thus, the
+	// swap() (close() + clearing of SubCons) operation is also guarded by this
+	// currentMu. Whether the current has been deleted or not will be checked
+	// from reading from the scToSubBalancer map, as that will be cleared in the
+	// atomic swap() operation. We are guaranteed that there can only be one
+	// balancer closed within this UpdateSubConnState() call, as multiple
+	// balancers closed would require another call to SwitchTo() which we are
+	// guaranteed won't be called concurrently. UpdateSubConnState() is
+	// different to UpdateClientConnState() and ResolverError(), as for those
+	// functions you don't update the current balancer in the situation where
+	// the current balancer could be closed (current + pending balancer
+	// populated).
+	gsb.currentMu.Lock()
+	defer gsb.currentMu.Unlock()
+	gsb.mu.Lock()
+	if gsb.closed {
+		gsb.mu.Unlock()
+		return
+	}
+	// This SubConn update will forward to the current balancer even if there is
+	// a pending present. This is because if this balancer does not forward the
+	// update, the picker from the current will not be updated with SubConns,
+	// leading to the possibility that the ClientConn constantly picks bad
+	// SubConns in the Graceful Switch period.
+	balToUpdate, ok := gsb.scToSubBalancer[sc]
+	if !ok {
+		gsb.mu.Unlock()
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		delete(gsb.scToSubBalancer, sc)
+	}
+	gsb.mu.Unlock()
+	balToUpdate.UpdateSubConnState(sc, state)
+}
+
+func (gsb *gracefulSwitchBalancer) Close() {
+	gsb.mu.Lock()
+	gsb.closed = true
+	for sc := range gsb.scToSubBalancer {
+		gsb.cc.RemoveSubConn(sc)
+		delete(gsb.scToSubBalancer, sc)
+	}
+	currentBalancerToUpdate := gsb.balancerCurrent
+	gsb.balancerCurrent = nil
+	pendingBalancerToUpdate := gsb.balancerPending
+	gsb.balancerPending = nil
+	gsb.mu.Unlock()
+
+	if currentBalancerToUpdate != nil {
+		currentBalancerToUpdate.Close()
+	}
+	if pendingBalancerToUpdate != nil {
+		pendingBalancerToUpdate.Close()
+	}
+}
+
+func (gsb *gracefulSwitchBalancer) ExitIdle() {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if gsb.balancerCurrent != nil {
+		if ei, ok := gsb.balancerCurrent.(balancer.ExitIdler); ok {
+			ei.ExitIdle()
+		}
+	}
+	if gsb.balancerPending != nil {
+		if ei, ok := gsb.balancerPending.(balancer.ExitIdler); ok {
+			ei.ExitIdle()
+		}
+	}
+}
+
+type clientConnWrapper struct {
+	gsb *gracefulSwitchBalancer
+
+	bal balancer.Balancer
```

Comment:
`bal` needs to be protected by the mutex

And the write at line 198

Reply:
And add a test (using the stub balancer: https://github.com/grpc/grpc-go/blob/master/internal/balancer/stub/stub.go#L30)

Reply:
It's worse than that.  :wink:  There are bigger changes coming